tests/test_target.py

```diff
@@ -37,23 +36,57 @@ def build_targets():
 
     targets = []
 
+    if target_configs.get('AndroidTarget') is not None:
+        print('> Android targets:')
+        for entry in target_configs['AndroidTarget'].values():
+            pp(entry)
+            a_target = AndroidTarget(
+                connection_settings=entry['connection_settings'],
+                conn_cls=lambda **kwargs: AdbConnection(adb_as_root=True, **kwargs),
+            )
+            targets.append((a_target, None))
+
+    if target_configs.get('LinuxTarget') is not None:
+        print('> Linux targets:')
+        for entry in target_configs['LinuxTarget'].values():
+            pp(entry)
+            l_target = LinuxTarget(connection_settings=entry['connection_settings'])
+            targets.append((l_target, None))
+
     if target_configs.get('LocalLinuxTarget') is not None:
         print('> LocalLinux targets:')
         for entry in target_configs['LocalLinuxTarget'].values():
             pp(entry)
             ll_target = LocalLinuxTarget(connection_settings=entry['connection_settings'])
-            targets.append(ll_target)
+            targets.append((ll_target, None))
```

Comment:
Returning `None` is a bad idea: https://github.com/ARM-software/devlib/pull/669/files#r1541156876
I do note that you avoided your own trap, so hats off :) , but I can guarantee you'll fall in it at one point or another without type checking (I've never seen anyone not fall in it, me included)

The cleanest approach here is to simply create a no-op target runner that implements the API and does nothing apart from exposing a `.target` attribute. This way, the rest of the code does not need to care about special cases, and if part of the API is impossible to implement, that will just have revealed a design problem that needs to be addressed.

Reply:
Done - kind of.
Let's continue discussion on the PR that I'm gonna file for addressing these comments.